### ISSUE TITLE
node_sqlite3 native module build issue

### ISSUE THREAD

**Issue by saaiful** [nexe/nexe#287]
I'm using sqlite3 in my application. I tried using nexe to compile my application and I get usual successful build, when I tried to run the application it throw the following error:
```
bootstrap_node.js:461
      throw new Error(`No such native module ${id}`);
      ^

Error: No such native module E:\pos\node_modules\sqlite3\package.json
    at NativeModule.require (bootstrap_node.js:461:13)
    at s (nexe.js:1:176)
    at nexe.js:1:367
    at Object.exports.find (nexe.js:19518:23)
    at Object.<anonymous> (nexe.js:19070:27)
    at Object.__dirname.66../trace (nexe.js:19257:4)
    at s (nexe.js:1:316)
    at nexe.js:1:367
    at Object.__dirname.1.express (nexe.js:3:15)
    at s (nexe.js:1:316)
```

After hours of search and reading, I found a solution where I have to manually copy some .cc/.h file in nexe's src directory. Now I'm getting the following error:
```
  cctest.vcxproj -> E:\pos\tmp\nexe\nodejs\latest\node-v7.2.1\Release\cctest.pdb (Full PDB)
  node_javascript.cc
     Creating library E:\pos\tmp\nexe\nodejs\latest\node-v7.2.1\Release\node.lib and object E:\pos\tmp\nexe\nodejs\latest\node-v7.2.1\Release\node.exp
node_sqlite3.obj : error LNK2001: unresolved external symbol "public: static void __cdecl node_sqlite3::Database::Init(class v8::Local<class v8::Object>)" (?Init@Database@node_sqlite3@@SAXV?$Local@VObject@v8@@@v8@@@Z) [E:\pos\tmp\nexe\nodejs\latest\node-v7.2.1\node.vcxproj]
node_sqlite3.obj : error LNK2001: unresolved external symbol "public: static void __cdecl node_sqlite3::Statement::Init(class v8::Local<class v8::Object>)" (?Init@Statement@node_sqlite3@@SAXV?$Local@VObject@v8@@@v8@@@Z) [E:\pos\tmp\nexe\nodejs\latest\node-v7.2.1\node.vcxproj]
E:\pos\tmp\nexe\nodejs\latest\node-v7.2.1\Release\node.exe : fatal error LNK1120: 2 unresolved externals [E:\pos\tmp\nexe\nodejs\latest\node-v7.2.1\node.vcxproj]
....> ERROR: The release executable has not been generated. This indicates a failure in the build process. There is likely additional information above.
```

What to do ? 

Btw, is there any way to use prebuild ".node" ?

**Comment by calebboyd** [comment]
The nexe bundler doesn't support native modules out of the box. However you can configure a webpack bundle to accomplish what you want.

I made an example [gist](https://gist.github.com/calebboyd/d1729d4c3903d10c88da292942ddc3a5) bundling an example project using `sqlite3` You can then use that bundle file as input

**Comment by calebboyd** [comment]
Note that the rewrite loader is hardcoded for win32-x64-6.9.2.  You'll have to use the same versions to build as you are running in your environment. See this [issue](https://github.com/calebboyd/xbin/issues/4) for more details

**Comment by jaredallard** [comment]
We'll hopefully streamline this!

**Comment by saaiful** [comment]
@calebboyd Thanks!
After following your instruction I'm getting the Following Error
```
nexe.js:306
            process.dlopen(module, path);
                    ^

Error: %1 is not a valid Win32 application.
E:\pos\5c69eab898b5ea956569347dc1bbca33.node
    at Object.<anonymous> (nexe.js:306:14)
    at Object.module.exports.module.deprecate (nexe.js:308:31)
    at __webpack_require__ (nexe.js:21:30)
    at webpackContext (nexe.js:284:10)
    at Object.<anonymous> (nexe.js:70:38)
    at __webpack_require__ (nexe.js:21:30)
    at Object.<anonymous> (nexe.js:48:16)
    at __webpack_require__ (nexe.js:21:30)
    at nexe.js:41:18
    at Object.__dirname.1.events (nexe.js:44:10)
```
My Sqlite Binary is "node-v51-win32-x64".

**Comment by calebboyd** [comment]
It sounds like your configuration is the following:

System Node version: 7.2.0 x64
Sqlite3 binary version: 7.2.0 x64
Nexe build version: 7.2.0 ia32  (32 bit is default)

You can either pass `nodeVCBuildArgs: ["nosign", "x64"]` to nexe or change your system version and rebuild sqlite3 with something like nvm `nvm install 7.2.0 32 && nvm use 7.2.0 32`

Cheers!

**Comment by mracsys** [comment]
I ran into the same problem and had trouble getting the x64 build argument to pass through to vcbuilt.bat. nodeVCBuildArgs inside nexe.compile() does not appear to work on v1.1.2. I resolved this by running nexe, deleting the extracted Node source folder, manually re-extracting it, and manually changing the following line in vcbuilt.bat:

`set target_arch=x86`
changed to
`set target_arch=x64`

After this change nexe was able to build a valid exe. Hopefully this helps someone else!

**Comment by calebboyd** [comment]
@saaiful Did you get everything building alright?

**Comment by calebboyd** [comment]
tracking in #396

**Comment by shirshak55** [comment]
@calebboyd is there no better way to do without using webpack etc?

**Comment by calebboyd** [comment]
@vercel/ncc which uses https://github.com/vercel/webpack-asset-relocator-loader
to rewrite locations for dlls (.node) files. Is the highest evolution to date of the approaches I've used. The implication is that your application reifies those .node files on first boot, so your code (with the rewritten paths to .node assets) can access them.

I maintained a similar asset rewriter with fusebox in nexe v2... but since dynamic linking from memory is not something that is straightforward to accomplish (at least from my perspective) the dll always needs to exist alone in/on _a_ kernel space filesystem. 

node-packer is a ruby project that achieves this by maintaining a fork of node and using squashfs. (afaik this is a huge maintenance burden)

facebookincubator/xar is another squashfs based "container of executables" tool. These both require squashfs to be available on the host though so are less practical except for specific usecases.
